File: test-scripts/create-test-pdf.js
/**
 * Create Test PDF
 * 
 * This script creates a simple test PDF for testing the FinDoc Analyzer application.
 */

const fs = require('fs');
const path = require('path');
const { PDFDocument, rgb, StandardFonts } = require('pdf-lib');

// Configuration
const config = {
  outputDir: path.join(__dirname, 'test-files'),
  fileName: 'simple-financial-statement.pdf'
};

// Create output directory if it doesn't exist
fs.mkdirSync(config.outputDir, { recursive: true });

/**
 * Create a simple financial statement PDF
 */
async function createSimpleFinancialStatementPDF() {
  console.log('Creating simple financial statement PDF...');
  
  // Create a new PDF document
  const pdfDoc = await PDFDocument.create();
  
  // Add a page
  const page = pdfDoc.addPage();
  
  // Get the standard font
  const font = await pdfDoc.embedFont(StandardFonts.Helvetica);
  const boldFont = await pdfDoc.embedFont(StandardFonts.HelveticaBold);
  
  // Set page dimensions
  const { width, height } = page.getSize();
  
  // Add title
  page.drawText('INVESTMENT PORTFOLIO STATEMENT', {
    x: 50,
    y: height - 50,
    size: 16,
    font: boldFont,
    color: rgb(0, 0, 0)
  });
  
  // Add header information
  page.drawText('Date: 31.03.2024', {
    x: 50,
    y: height - 80,
    size: 10,
    font: font,
    color: rgb(0, 0, 0)
  });
  
  page.drawText('Account Number: 87654321', {
    x: 50,
    y: height - 100,
    size: 10,
    font: font,
    color: rgb(0, 0, 0)
  });
  
  page.drawText('Client: Jane Smith', {
    x: 50,
    y: height - 120,
    size: 10,
    font: font,
    color: rgb(0, 0, 0)
  });
  
  // Add portfolio summary
  page.drawText('PORTFOLIO SUMMARY', {
    x: 50,
    y: height - 150,
    size: 14,
    font: boldFont,
    color: rgb(0, 0, 0)
  });
  
  page.drawText('Total Value: USD 875,000.00', {
    x: 50,
    y: height - 170,
    size: 10,
    font: font,
    color: rgb(0, 0, 0)
  });
  
  page.drawText('Currency: USD', {
    x: 50,
    y: height - 190,
    size: 10,
    font: font,
    color: rgb(0, 0, 0)
  });
  
  page.drawText('Valuation Date: 31.03.2024', {
    x: 50,
    y: height - 210,
    size: 10,
    font: font,
    color: rgb(0, 0, 0)
  });
  
  // Add asset allocation
  page.drawText('ASSET ALLOCATION', {
    x: 50,
    y: height - 240,
    size: 14,
    font: boldFont,
    color: rgb(0, 0, 0)
  });
  
  // Draw asset allocation table
  const assetAllocationData = [
    ['Asset Class', 'Percentage'],
    ['Equity', '55%'],
    ['Fixed Income', '25%'],
    ['Cash', '10%'],
    ['Alternative', '10%']
  ];
  
  drawTable(page, assetAllocationData, 50, height - 260, 200, 20, font, boldFont);
  
  // Add securities holdings
  page.drawText('SECURITIES HOLDINGS', {
    x: 50,
    y: height - 380,
    size: 14,
    font: boldFont,
    color: rgb(0, 0, 0)
  });
  
  // Draw securities table
  const securitiesData = [
    ['Security', 'ISIN', 'Type', 'Quantity', 'Price', 'Value', 'Weight'],
    ['APPLE INC', 'US0378331005', 'Equity', '200', 'USD 175.50', 'USD 35,100.00', '4.01%'],
    ['MICROSOFT CORP', 'US5949181045', 'Equity', '150', 'USD 420.00', 'USD 63,000.00', '7.20%'],
    ['AMAZON.COM INC', 'US0231351067', 'Equity', '75', 'USD 180.25', 'USD 13,518.75', '1.55%'],
    ['NVIDIA CORP', 'US67066G1040', 'Equity', '100', 'USD 950.00', 'USD 95,000.00', '10.86%'],
    ['US TREASURY 2.0% 15/02/2050', 'US912810SQ26', 'Bond', '150,000', 'USD 0.92', 'USD 138,000.00', '15.77%']
  ];
  
  drawTable(page, securitiesData, 50, height - 400, 70, 20, font, boldFont);
  
  // Serialize the PDF to bytes
  const pdfBytes = await pdfDoc.save();
  
  // Write to file
  const outputPath = path.join(config.outputDir, config.fileName);
  fs.writeFileSync(outputPath, pdfBytes);
  
  console.log(`PDF created at: ${outputPath}`);
  
  return outputPath;
}

/**
 * Draw a table on a PDF page
 * @param {PDFPage} page - PDF page
 * @param {Array<Array<string>>} data - Table data
 * @param {number} x - X position
 * @param {number} y - Y position
 * @param {number} cellWidth - Cell width
 * @param {number} cellHeight - Cell height
 * @param {PDFFont} font - Regular font
 * @param {PDFFont} headerFont - Header font
 */
function drawTable(page, data, x, y, cellWidth, cellHeight, font, headerFont) {
  const numColumns = data[0].length;
  const numRows = data.length;
  
  // Draw table grid
  for (let i = 0; i <= numRows; i++) {
    page.drawLine({
      start: { x, y: y - i * cellHeight },
      end: { x: x + numColumns * cellWidth, y: y - i * cellHeight },
      thickness: 1,
      color: rgb(0, 0, 0)
    });
  }
  
  for (let i = 0; i <= numColumns; i++) {
    page.drawLine({
      start: { x: x + i * cellWidth, y },
      end: { x: x + i * cellWidth, y: y - numRows * cellHeight },
      thickness: 1,
      color: rgb(0, 0, 0)
    });
  }
  
  // Draw table content
  for (let i = 0; i < numRows; i++) {
    for (let j = 0; j < numColumns; j++) {
      const cellFont = i === 0 ? headerFont : font;
      
      page.drawText(data[i][j], {
        x: x + j * cellWidth + 5,
        y: y - i * cellHeight - 15,
        size: 8,
        font: cellFont,
        color: rgb(0, 0, 0)
      });
    }
  }
}

// Create the PDF if this script is executed directly
if (require.main === module) {
  createSimpleFinancialStatementPDF().catch(console.error);
}

module.exports = createSimpleFinancialStatementPDF;

File: create_test_pdf.py
"""
Create Test PDF

This script creates a test PDF with financial data for testing.
"""

import os
from reportlab.lib.pagesizes import letter
from reportlab.platypus import SimpleDocTemplate, Paragraph, Spacer, Table, TableStyle
from reportlab.lib.styles import getSampleStyleSheet, ParagraphStyle
from reportlab.lib.enums import TA_LEFT, TA_CENTER, TA_RIGHT
from reportlab.lib import colors

def create_test_pdf(output_path):
    """
    Create a test PDF with financial data.
    
    Args:
        output_path: Output file path
    """
    # Create PDF
    doc = SimpleDocTemplate(output_path, pagesize=letter)
    styles = getSampleStyleSheet()
    
    # Create custom styles
    title_style = ParagraphStyle(
        'Title',
        parent=styles['Title'],
        fontSize=16,
        alignment=TA_CENTER,
        spaceAfter=12
    )
    
    heading_style = ParagraphStyle(
        'Heading',
        parent=styles['Heading2'],
        fontSize=14,
        alignment=TA_LEFT,
        spaceAfter=6,
        spaceBefore=12
    )
    
    subheading_style = ParagraphStyle(
        'Subheading',
        parent=styles['Heading3'],
        fontSize=12,
        alignment=TA_LEFT,
        spaceAfter=6
    )
    
    normal_style = ParagraphStyle(
        'Normal',
        parent=styles['Normal'],
        fontSize=10,
        alignment=TA_LEFT,
        spaceAfter=6
    )
    
    # Create content
    content = []
    
    # Title
    content.append(Paragraph("INVESTMENT PORTFOLIO STATEMENT", title_style))
    content.append(Spacer(1, 12))
    
    # Header information
    content.append(Paragraph("Date: 28.02.2025", normal_style))
    content.append(Paragraph("Account Number: 12345678", normal_style))
    content.append(Paragraph("Client: John Doe", normal_style))
    content.append(Spacer(1, 12))
    
    # Portfolio summary
    content.append(Paragraph("PORTFOLIO SUMMARY", heading_style))
    content.append(Paragraph("Total Value: USD 1,250,000.00", normal_style))
    content.append(Paragraph("Currency: USD", normal_style))
    content.append(Paragraph("Valuation Date: 28.02.2025", normal_style))
    content.append(Spacer(1, 12))
    
    # Asset allocation
    content.append(Paragraph("ASSET ALLOCATION", heading_style))
    
    asset_allocation_data = [
        ["Asset Class", "Percentage"],
        ["Equity", "45%"],
        ["Fixed Income", "30%"],
        ["Cash", "15%"],
        ["Alternative", "10%"]
    ]
    
    asset_allocation_table = Table(asset_allocation_data, colWidths=[200, 100])
    asset_allocation_table.setStyle(TableStyle([
        ('BACKGROUND', (0, 0), (-1, 0), colors.lightgrey),
        ('TEXTCOLOR', (0, 0), (-1, 0), colors.black),
        ('ALIGN', (0, 0), (-1, 0), 'CENTER'),
        ('FONTNAME', (0, 0), (-1, 0), 'Helvetica-Bold'),
        ('FONTSIZE', (0, 0), (-1, 0), 10),
        ('BOTTOMPADDING', (0, 0), (-1, 0), 12),
        ('BACKGROUND', (0, 1), (-1, -1), colors.white),
        ('TEXTCOLOR', (0, 1), (-1, -1), colors.black),
        ('ALIGN', (0, 1), (0, -1), 'LEFT'),
        ('ALIGN', (1, 1), (1, -1), 'RIGHT'),
        ('FONTNAME', (0, 1), (-1, -1), 'Helvetica'),
        ('FONTSIZE', (0, 1), (-1, -1), 10),
        ('GRID', (0, 0), (-1, -1), 1, colors.black),
        ('VALIGN', (0, 0), (-1, -1), 'MIDDLE'),
    ]))
    
    content.append(asset_allocation_table)
    content.append(Spacer(1, 12))
    
    # Securities holdings
    content.append(Paragraph("SECURITIES HOLDINGS", heading_style))
    
    securities_data = [
        ["Security", "ISIN", "Type", "Quantity", "Price", "Value", "Weight"],
        ["APPLE INC", "US0378331005", "Equity", "500", "USD 170.00", "USD 85,000.00", "6.8%"],
        ["MICROSOFT CORP", "US5949181045", "Equity", "300", "USD 340.00", "USD 102,000.00", "8.16%"],
        ["AMAZON.COM INC", "US0231351067", "Equity", "100", "USD 950.00", "USD 95,000.00", "7.6%"],
        ["US TREASURY 2.5% 15/02/2045", "US912810RK35", "Bond", "200,000", "USD 0.99", "USD 198,000.00", "15.84%"],
        ["GOLDMAN SACHS 0% NOTES 23-07.11.29", "XS2692298537", "Bond", "150,000", "USD 0.98", "USD 147,000.00", "11.76%"]
    ]
    
    securities_table = Table(securities_data, colWidths=[120, 100, 60, 60, 80, 100, 60])
    securities_table.setStyle(TableStyle([
        ('BACKGROUND', (0, 0), (-1, 0), colors.lightgrey),
        ('TEXTCOLOR', (0, 0), (-1, 0), colors.black),
        ('ALIGN', (0, 0), (-1, 0), 'CENTER'),
        ('FONTNAME', (0, 0), (-1, 0), 'Helvetica-Bold'),
        ('FONTSIZE', (0, 0), (-1, 0), 10),
        ('BOTTOMPADDING', (0, 0), (-1, 0), 12),
        ('BACKGROUND', (0, 1), (-1, -1), colors.white),
        ('TEXTCOLOR', (0, 1), (-1, -1), colors.black),
        ('ALIGN', (0, 1), (2, -1), 'LEFT'),
        ('ALIGN', (3, 1), (-1, -1), 'RIGHT'),
        ('FONTNAME', (0, 1), (-1, -1), 'Helvetica'),
        ('FONTSIZE', (0, 1), (-1, -1), 8),
        ('GRID', (0, 0), (-1, -1), 1, colors.black),
        ('VALIGN', (0, 0), (-1, -1), 'MIDDLE'),
    ]))
    
    content.append(securities_table)
    content.append(Spacer(1, 12))
    
    # Sector allocation
    content.append(Paragraph("SECTOR ALLOCATION", heading_style))
    
    sector_allocation_data = [
        ["Sector", "Percentage"],
        ["Technology", "22.56%"],
        ["Consumer", "7.6%"],
        ["Government", "15.84%"],
        ["Financial", "11.76%"],
        ["Other", "42.24%"]
    ]
    
    sector_allocation_table = Table(sector_allocation_data, colWidths=[200, 100])
    sector_allocation_table.setStyle(TableStyle([
        ('BACKGROUND', (0, 0), (-1, 0), colors.lightgrey),
        ('TEXTCOLOR', (0, 0), (-1, 0), colors.black),
        ('ALIGN', (0, 0), (-1, 0), 'CENTER'),
        ('FONTNAME', (0, 0), (-1, 0), 'Helvetica-Bold'),
        ('FONTSIZE', (0, 0), (-1, 0), 10),
        ('BOTTOMPADDING', (0, 0), (-1, 0), 12),
        ('BACKGROUND', (0, 1), (-1, -1), colors.white),
        ('TEXTCOLOR', (0, 1), (-1, -1), colors.black),
        ('ALIGN', (0, 1), (0, -1), 'LEFT'),
        ('ALIGN', (1, 1), (1, -1), 'RIGHT'),
        ('FONTNAME', (0, 1), (-1, -1), 'Helvetica'),
        ('FONTSIZE', (0, 1), (-1, -1), 10),
        ('GRID', (0, 0), (-1, -1), 1, colors.black),
        ('VALIGN', (0, 0), (-1, -1), 'MIDDLE'),
    ]))
    
    content.append(sector_allocation_table)
    content.append(Spacer(1, 12))
    
    # Notes
    content.append(Paragraph("NOTES", heading_style))
    content.append(Paragraph("This portfolio statement is for informational purposes only and does not constitute investment advice. Past performance is not indicative of future results. Please consult with your financial advisor before making any investment decisions.", normal_style))
    
    # Build PDF
    doc.build(content)
    
    print(f"Test PDF created at {output_path}")

if __name__ == "__main__":
    output_dir = "test_pdfs"
    os.makedirs(output_dir, exist_ok=True)
    output_path = os.path.join(output_dir, "test_portfolio_statement.pdf")
    create_test_pdf(output_path)
